File: day14/day14.kt
import java.io.File

val toBin = { v: Long -> java.lang.Long.toBinaryString(v).padStart(36, '0') }
data class SetMask(val mask: String)
data class WriteValue(val addr: Long, val value: Long)

fun main() {
    val program = File("day14.txt").readLines().map {
        val m = Regex("""mask = ([X01]+)|mem\[(\d+)\] = (\d+)""").find(it)!!.groups
        if (m[1] != null) SetMask(m[1]!!.value) else
            WriteValue(m[2]!!.value.toLong(), m[3]!!.value.toLong())
    }

    val memory = mutableMapOf<Long,Long>()
    var mask = ""
    program.forEach {
        when {
            it is SetMask -> mask = it.mask
            it is WriteValue -> {
                memory[it.addr] = mask.zip(toBin(it.value)) {
                    a, b -> if (a == 'X') b else a
                }.joinToString("").toLong(2)
            }
        }
    }
    println(memory.values.sum())

    memory.clear()
    fun write(addr: String, value: Long) {
        if (addr.contains('X')) {
            write(addr.replaceFirst('X', '0'), value)
            write(addr.replaceFirst('X', '1'), value)
        } else memory[addr.toLong(2)] = value
    }
    program.forEach {
        when {
            it is SetMask -> mask = it.mask
            it is WriteValue -> {
                write(mask.zip(toBin(it.addr)) {
                    a, b -> if (a == '0') b else a
                }.joinToString(""), it.value)
            }
        }
    }
    println(memory.values.sum())
}

File: 2020/day14/day14.py
import re

program = []
for line in open("day14.txt"):
    m = re.match(r'mask = ([X01]+)|mem\[(\d+)\] = (\d+)', line)
    program.append(m.groups())

memory, or_mask, and_mask = {}, None, None
for mask, addr, value in program:
    if mask:
        and_mask = int(''.join('1' if c == 'X' else '0' for c in mask), 2)
        or_mask = int(mask.replace('X', '0'), 2)
    elif addr:
        memory[int(addr)] = (int(value) & and_mask) | or_mask
print(sum(memory.values()))

memory, addr_mask = {}, None
def write(addr, value):
    if 'X' in addr:
        write(addr.replace('X', '0', 1), value)
        write(addr.replace('X', '1', 1), value)
    else:
        memory[int(addr, 2)] = value
for mask, addr, value in program:
    if mask:
        addr_mask = mask
    elif addr:
        addr_1 = bin(int(addr))[2:].rjust(len(addr_mask), '0')
        addr_2 = ''.join(addr_1[k] if v == '0' else v for k, v in enumerate(addr_mask))
        write(addr_2, int(value))
print(sum(memory.values()))
